src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -211,6 +198,41 @@ private async Task ProcessQueueAsync()
             }
         }
 
+        private async Task ProcessSingleWorkItemAsync(QueueItem work, int requestId)
+        {
+            var logIdName = $"{work.ClientName ?? "Default"}.{requestId++:00000000}.{work.MethodName}.{DateTime.Now:yyyy-MM-dd-HH-mm-ss}";
+            var traceSource = _logger == null ? null : await _logger.CreateTraceSourceAsync(logIdName, _cancelSource.Token).ConfigureAwait(false);
+
+            try
+            {
+                // Create a linked cancellation token to cancel any requests in progress when this shuts down
+                var cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(_cancelSource.Token, work.CancellationToken).Token;
+
+                var context = CreateRequestContext(work, traceSource);
+
+                if (work.MutatesSolutionState)
+                {
+                    // Mutating requests block other requests from starting to ensure an up to date snapshot is used.
+                    await work.CallbackAsync(context, cancellationToken).ConfigureAwait(false);
+
+                    // Now that we've mutated our solution, clear out our saved state to ensure it gets recalculated
+                    _lspSolutionCache.Remove(context.Solution.Workspace);
+                }
+                else
+                {
+                    // Non mutating are fire-and-forget because they are by definition readonly. Any errors
+                    // will be sent back to the client but we can still capture errors in queue processing
+                    // via NFW, though these errors don't put us into a bad state as far as the rest of the queue goes.
+                    _ = work.CallbackAsync(context, cancellationToken).ReportNonFatalErrorAsync();
+                }
+            }
+            finally
+            {
+                traceSource?.Flush();
```

Comment:
For non-mutating requests, we don't block (line 226) for the request to finish, so this will flush and close the trace source before the request has been processed, potentially even before it starts processing.

If we're going to use a different trace source per request, the code to create and close it would need to go into the callback.

Reply:
ah ok.